peps/pep-0740.rst

```diff
@@ -205,110 +181,340 @@ changes to it:
+
+  raw_signature = signing_key.sign(encoded_payload, ECDSA(SHA2_256()))
+  message_signature = b64encode(raw_signature)
 
 .. _provenance-object:
 
 Provenance objects
-^^^^^^^^^^^^^^^^^^
+------------------
 
 The index will serve uploaded attestations along with metadata that can assist
 in verifying them in the form of JSON serialized objects.
 
-These "provenance objects" will be available via both the :pep:`503` Simple Index
-and :pep:`691` JSON-based Simple API as described below, and will have the
-following structure:
+These *provenance objects* will be available via both the :pep:`503` Simple Index
+and :pep:`691` JSON-based Simple API as described above, and will have the
+following layout:
 
 .. code-block:: json
 
     {
-        "publisher": {
-          "type": "important-ci-service",
-          "claims": {},
-          "vendor-property": "foo",
-          "another-property": 123
-        },
-        "attestations": {
-          "some-attestation": {/* ... */},
-          "another-attestation": {/* ... */}
-        }
+        "version": 1,
+        "attestations": [
+          [
+            {
+              "kind": "important-ci-service",
+              "claims": {},
+              "vendor-property": "foo",
+              "another-property": 123
+            },
+            [
+              { /* attestation 1 ... */ },
+              { /* attestation 2 ... */ }
+            ]
+          ]
+        ]
     }
 
-* ``publisher`` is an **optional** JSON object, containing a
-  representation of the file's Trusted Publisher configuration at the time
-  the file was uploaded to the package index. The keys within the ``publisher``
-  object are specific to each Trusted Publisher but include, at minimum:
-
-  * A ``type`` key, which **MUST** be a JSON string that uniquely identifies the
+or, as pseudocode:
+
+.. code-block:: python
+
+  @dataclass
+  class Publisher:
+      kind: string
+      """
+      The kind of Trusted Publisher.
+      """
+
+      claims: object | None
+      """
+      Any context-specific claims retained by the index during Trusted Publisher
+      authentication.
+      """
+
+      _rest: object
+      """
+      Each publisher object is open-ended, meaning that it MAY contain additional
+      fields beyond the ones specified explicitly above. This field signals that,
+      but is not itself present.
+      """
+
+  @dataclass
+  class Provenance:
+      version: Literal[1]
+      """
+      The provenance object's version, which is always 1.
+      """
+
+      attestations: list[tuple[Publisher, list[Attestation]]]
+      """
+      One or more attestation objects, indexed by their publisher.
+      """
+
+* ``version`` is ``1``. Like attestation objects, provenance objects are
+  versioned, and this PEP only defines version ``1``.
+* ``attestations`` is a **required** JSON array, containing pairs
+  of publishers and a JSON array of one or more attestation objects.
```

Comment:
Why do we use a tuple of `(publisher, List[attestations])`, and not an actual object like:
```JSON
{
    "publisher": {....}
    "attestations": [...]
}
```
(I'm mostly thinking that the object would be easier to extend in the future, rather than a tuple where the meaning is determined by the element order)

Reply:
No good reason -- your suggestion is much cleaner 🙂.

How does this sound?

```
class PublisherAttestations:
  publisher: Publisher
  attestations: list[Attestation]

class Provenance:
  version: Literal[1]
  attestations: list[PublisherAttestations]
```

(The `PublisherAttestations` name might need workshopping.)

Reply:
Yeah. I also think we should rename the `attestations` field in `Provenance`, since it's not _only_ providing attestations. And with the current naming, you'd have things like:
```python
provenance_object['attestations'][0]['attestations'][0]
```
due to `attestations` being the name of the field in both structs.

Since what we want to encode is that each Attestation object is associated with some metadata needed to verify it (in this case, a `Publisher`), we could have something like:

```
class AttestationBundle:
  publisher: Publisher
  attestations: list[Attestation]

class Provenance:
  version: Literal[1]
  attestation_bundles: list[AttestationBundle]
```

In this case, `AttestationBundle` is a bundle of `Attestations` + some metadata related to them. But yeah, I'm also open to better naming suggestions.

Reply:
SGTM. `attestation_bundles` is much better than overloading `attestations`, and we can iterate if a better name comes along 🙂